src/docGenerator.ts

```diff
@@ -0,0 +1,132 @@
+import { JSONSchema4 } from "json-schema"
+import { flatMap, flatMapDeep } from "lodash"
+import fetch from "node-fetch"
+import { DescriptionEntry } from "./model/Description"
+import {
+  Category,
+  fromEslintCategoryToCategory,
+  fromEslintCategoryToLevel,
+  Level,
+  patternIdToCodacy,
+  Patterns,
+  PatternsEntry,
+  PatternsParameter
+} from "./model/Patterns"
+import { fromSchemaArray } from "./NamedParameters"
+import { toolName, toolVersion } from "./toolMetadata"
+import { writeFile } from "./fileUtils"
+import { capitalize, patternTitle } from "./DocGeneratorStringUtils"
+import { Rule } from "eslint"
+export class DocGenerator {
+  private readonly rules: Map<string, Rule.RuleModule>
+
+  constructor(rules: Map<string, Rule.RuleModule>) {
+    this.rules = rules
+  }
+
+  generatePatterns(): Patterns {
+    const entries = flatMap(
+      Array.from(this.rules.entries()),
+      ([patternId, ruleModule]) => {
+        const meta = ruleModule && ruleModule.meta ? ruleModule.meta : undefined
+        const eslintCategory =
+          meta && meta.docs ? meta.docs.category : undefined
+        const level: Level = fromEslintCategoryToLevel(eslintCategory)
+        const category: Category = patternId.includes("security")
+          ? "Security"
+          : fromEslintCategoryToCategory(eslintCategory)
+        const parameters =
+          meta && meta.schema
+            ? this.fromEslintSchemaToParameters(meta.schema)
+            : undefined
+        return new PatternsEntry(
+          patternIdToCodacy(patternId),
+          level,
+          category,
+          parameters && parameters.length > 0 ? parameters : undefined
+        )
+      }
+    )
+    return new Patterns(toolName, toolVersion, entries)
+  }
+
+  generateDescription(): DescriptionEntry[] {
+    const descriptionEntries = Array.from(this.rules.entries()).map(
+      ([patternId, ruleModule]) => {
+        const eslintDescription =
+          ruleModule && ruleModule.meta && ruleModule.meta.docs
+            ? ruleModule.meta.docs.description
+            : undefined
+        const description = eslintDescription
+          ? capitalize(eslintDescription)
+          : ""
+        const title = patternTitle(patternId)
+        const timeToFix = 5
+        return new DescriptionEntry(
+          patternIdToCodacy(patternId),
+          title,
+          description,
+          timeToFix
+        )
+      }
+    )
+    return descriptionEntries.filter(x => x !== null) as DescriptionEntry[]
+  }
+
+  private fromEslintSchemaToParameters(
+    schema: JSONSchema4 | JSONSchema4[]
+  ): PatternsParameter[] {
+    const anyOfToArray = (schema: JSONSchema4) => {
+      return schema.anyOf ? schema.anyOf : [schema]
```

Comment:
I don't like the position of this `return`
and is quite hard to follow all of the exit conditions of this method

Reply:
✅Fixed